### PR TITLE
pdm: 2.10.4 -> 2.12.1, fix build

### LINKED ISSUE #281289: Build failure: `pdm`
### Steps To Reproduce

1. build *truststore*

### Build log

Output of `nix log github:NixOS/nixpkgs#pdm`:

```
...
 - Adding pdm-2.10.4.dist-info/RECORD -> pdm-2.10.4.dist-info/RECORD
Successfully built pdm-2.10.4-py3-none-any.whl
Finished creating a wheel...
Finished executing pypaBuildPhase
@nix { "action": "setPhase", "phase": "pythonRuntimeDepsCheckHook" }
Running phase: pythonRuntimeDepsCheckHook
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for pdm-2.10.4-py3-none-any.whl
  - truststore not installed
```

### Additional context

Adding `python3.pkgs.truststore` to `propagatedBuildInputs` solves this issue.

### Notify maintainers

@cpcloud 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.133.1-microsoft-standard-WSL2, NixOS, 24.05 (Uakari), 24.05.20240106.46ae021`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/281289

This PR adds `dep-logic`, which is a new dependency of `pdm`.
I also added the `truststore` python package as a dependency, which was missing, and caused a build failure.

I also switched over to `pyproject = true` instead of the legacy way of doing it.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/dep-logic/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  src = fetchFromGitHub {
+    owner = "pdm-project";
+    repo = "dep-logic";
+    rev = version;
+    hash = "sha256-AFiCNzHlo3BADqbjRBruA80cfM6Ytdb+gReEg5hUmro=";
+  };
+
+  nativeBuildInputs = [ pdm-backend ];
+
```
Comment (dotlambda) [inline pkgs/development/python-modules/dep-logic/default.nix:24]:
```suggestion

propagatedBuildInputs = [
  packaging
];

```

**Review by dotlambda — COMMENTED** [review]
2.12.0 is out

File: pkgs/development/python-modules/dep-logic/default.nix (hunk 1 above)
Comment (TomaSajt) [reply pkgs/development/python-modules/dep-logic/default.nix:24]: About the `packaging` dependency:
it seemed to build OK without it somehow and then I forgot to add it. Should just every package listed as a dependency always be included?

Comment (dotlambda) [reply pkgs/development/python-modules/dep-logic/default.nix:27]: > Should just every package listed as a dependency always be included?

Yes.

**Review by dotlambda — APPROVED** [review]

**Comment by XYenon** [comment]
Result of `nixpkgs-review pr 281328` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>pdm</li>
    <li>pdm.dist</li>
    <li>python311Packages.dep-logic</li>
    <li>python311Packages.dep-logic.dist</li>
    <li>python312Packages.dep-logic</li>
    <li>python312Packages.dep-logic.dist</li>
  </ul>
</details>

**Review by XYenon — APPROVED** [review]

**Pull request merged**